src/Mutant/MutationFactory.php

```diff
@@ -66,35 +74,103 @@ public function __construct(
         $this->tmpDir = $tmpDir;
         $this->differ = $differ;
         $this->printer = $printer;
-        $this->mutantCodeFactory = $mutantCodeFactory;
+        $this->codeFactory = $mutantCodeFactory;
     }
 
-    public function create(Mutation $mutation): Mutant
-    {
-        $mutantFilePath = sprintf(
-            '%s/mutant.%s.infection.php',
+    /**
+     * @param Node[] $originalFileAst
+     * @param array<string|int|float> $attributes
+     * @param class-string $mutatedNodeClass
+     * @param TestLocation[] $tests
+     */
+    public function create(
+        string $originalFilePath,
+        array $originalFileAst,
+        string $mutatorName,
+        array $attributes,
+        string $mutatedNodeClass,
+        MutatedNode $mutatedNode,
+        int $mutationByMutatorIndex,
+        array $tests
+    ): Mutation {
+        foreach (MutationAttributeKeys::ALL as $key) {
+            Assert::keyExists($attributes, $key);
+        }
+
+        $attributes = array_intersect_key($attributes, array_flip(MutationAttributeKeys::ALL));
+
+        $hash = self::createHash(
+            $originalFilePath,
+            $mutatorName,
+            $attributes,
+            $mutationByMutatorIndex
+        );
+
+        $mutationFilePath = sprintf(
+            '%s/mutation.%s.infection.php',
```

Comment:
I have to say I'm lost here. 

Even from the nomenclature added in this PR, Mutation is:

> The result of applying a mutator **to the AST** of a subject and represents **a change to be applied**.

But the result source code after applying a _Mutation_ is called **Mutant**. 

> New program that differs from the original by applying a mutation.

IMO this conflicts with definitions from Mutation Testing theory (not even in Infection in particular)

---

After more digging into this patch, here is what we do in this PR:

```php
$mutatedCode = $codeFactory->createCode();

# later in `MutationTestingRunner.php` we dump it to FS

$filesystem->dumpFile('mutation.%s.infection.php', $mutatedCode)
```

How it was before (from my POV - much better)

```php
$mutatedCode = $mutantCodeFactory->createCode();

# later in `MutationTestingRunner.php` we dump it to FS

$filesystem->dumpFile('mutant.%s.infection.php', $mutatedCode)
```

Let's also see the name of `$mutationFilePath` variable. How can Mutation have a file path? From nomenclature, Mutation _"... represents a change to be applied"_, while Mutant is _"New program that differs from the original by applying a mutation."_

New program can have a file (where this program is located).
Change to be applied - not. (Mutation is _"The result of applying a mutator to the AST of a subject"_).

We store AST in memory, and we dumpt the result of mutated AST (which is a Mutant) to a file.

Reply:
I'm really thinking hard on this, so let's see if that makes sense.

The way I see it, the changed AST, a string diff, or a dumped file, is not the mutant, but the mutation.

I interpret _"New program that differs from the original by applying a mutation."_ more as "the process with the mutated code loaded", i.e. in our case, the "process in which we intercept the autoloader to load the mutated code instead of the original".

Maybe this is too much of a stretch? But honestly this makes more sense to me and also has another merit: it is not specific to PHP/Infection. Dumping the mutated code to the FS is a Infection specificity due to how PHP works, I don't think calling _that_ a mutant is correct